storage: streaming snapshots v2

jordanlewis opened the pull request:
This is a rework of the previous streaming snapshots work from #8726 for
compatibility with the removal of processRaftMu.

This commit changes the mechanism of transport for raft snapshots to a
new streaming GRPC method, as per the first phase of the streaming
snapshots RFC [0](https://github.com/cockroachdb/cockroach/blob/3bc6103fc46665b7089e4d80d1966f8f3e2bcbb2/docs/RFCS/streaming_snapshots.md). Snapshots are not yet truly streamed - this commit is
mostly a plumbing change. Instead, they are still sent in a large blob,
but over the new streaming transport.

Also, delete the unused `RaftMessageSync` method and associated
infrastructure as it is superseded by streaming snapshots.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9292)

<!-- Reviewable:end -->

jordanlewis committed:
storage/engine: improve batch.Repr()

If the batch has never been flushed, we can just return the mutations as
a Repr without having to flush to C++ and read bytes back.

(cherry picked from commit e606afe97e564e350ee59283410f11f5a9d2d936)

jordanlewis committed:
storage: stream snapshots via a new GRPC endpoint.

This is a rework of the previous streaming snapshots work from #8726 for
compatibility with the removal of processRaftMu.

This commit changes the mechanism of transport for raft snapshots to a
new streaming GRPC method, as per the first phase of the streaming
snapshots RFC [0]. Snapshots are not yet truly streamed - this commit is
mostly a plumbing change. Instead, they are still sent in a large blob,
but over the new streaming transport.

Also, delete the unused `RaftMessageSync` method and associated
infrastructure as it is superseded by streaming snapshots.

[0]: https://github.com/cockroachdb/cockroach/blob/3bc6103fc46665b7089e4d80d1966f8f3e2bcbb2/docs/RFCS/streaming_snapshots.md

jordanlewis commented:
Review status: 0 of 24 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/replica_raftstorage.go, line 280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRUTQCazzfm332k4CXU:-KRUTQCazzfm332k4CXV:b-ye4zum) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica_raftstorage.go#L280)):*
> ```Go
> // SnapshotWithContext is main implementation for Snapshot() but it takes a
> // context to allow tracing. If this method returns without error, callers
> // must eventually call CloseOutSnap to ready this replica for more snapshots.
> ```

Pasting unresolved comment from @bdarnell: This method is called by Replica.Snapshot, which is called by raft, which knows nothing of this requirement. It looks like currently any time raft calls this method we'll see it when we try to send a message from handleRaftReady, but I'm uneasy about relying on this property. I think that handleRaftReady should check to see if there is a pending snapshot, and if there was one but we didn't send a snapshot message, close it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r1, 21 of 21 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1670 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRXRT71Y5us3jnM_-hT:-KRXRT72rH9_MKo-p_4l:bo4ictp) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica.go#L1670)):*
> ```Go
> // handleRaftReady processes a raft.Ready containing entries and messages that
> // are ready to read, be saved to stable storage, committed or sent to other
> // peers. It takes a non-nil IncomingSnapshot pointer to indicate that it is
> ```

IncomingSnapshot is no longer a pointer.

---

*[storage/replica.go, line 1736 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRXRZnw0bEvT7EWSU6w:-KRXRZnw0bEvT7EWSU6x:b-w7wkwe) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica.go#L1736)):*
> ```Go
> 		}
> 		if *snapUUID != inSnap.SnapUUID {
> 			log.Fatalf(ctx, "programming error: a snapshot application was attempted outside of the streaming snapshot codepath")
> ```

It may be helpful to distinguish cases where inSnap.SnapUUID is empty versus cases where it has a value that doesn't match.

---

*[storage/replica_raftstorage.go, line 280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRUTQCazzfm332k4CXU:-KRXZmVg-Sauqe4srKZe:b9l9w3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica_raftstorage.go#L280)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> Pasting unresolved comment from @bdarnell: This method is called by Replica.Snapshot, which is called by raft, which knows nothing of this requirement. It looks like currently any time raft calls this method we'll see it when we try to send a message from handleRaftReady, but I'm uneasy about relying on this property. I think that handleRaftReady should check to see if there is a pending snapshot, and if there was one but we didn't send a snapshot message, close it.
</details>
My concern here is just the indirect nature of the path from here to CloseOutSnap. If raft fetches a snapshot from the replica but then for any reason fails to send the MsgSnap, we'll never close the snapshot and be blocked from sending any more snapshots of this replica in the future. Currently, it doesn't look like that can ever happen, but it also doesn't look like a very strong guarantee. I'd like to have handleRaftReady close any outgoing snapshot (not just when an MsgSnap is sent) as a safety net. 

---

*[storage/replica_raftstorage.go, line 353 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRXXbDCTGabKPLfWXC5:-KRXXbDDia_aJQogAivc:b-11z90f) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica_raftstorage.go#L353)):*
> ```Go
> 
> // OutgoingSnapshot contains the data required to stream a snapshot to a recipient. Once one is
> // created, it needs to be closed via Close() to prevent resource leakage.
> ```

s/Close/Replica.CloseOutSnap/

---

*[storage/store.go, line 2635 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRXUKPLviT9zqcXa3yj:-KRXUKPMcabwBK5sZBr0:bpmt2by) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/store.go#L2635)):*
> ```Go
> 			header.RangeDescriptor.RangeID, resp.Message)
> 	case SnapshotResponse_ACCEPTED:
> 	// continue
> ```

This looks like a commented-out `continue` statement, which wouldn't make since because there is no loop. Either remove it or make it a sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-:-KRXYJpTQ7GrRNd-6gts:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 1670 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRXRT71Y5us3jnM_-hT:-KRZ5ZJjkMi83CUD_lmm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica.go#L1670)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> IncomingSnapshot is no longer a pointer.
</details>
Done.

---

*[storage/replica.go, line 1736 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRXRZnw0bEvT7EWSU6w:-KRZ8KyedS_hD_ZKRkVH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica.go#L1736)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It may be helpful to distinguish cases where inSnap.SnapUUID is empty versus cases where it has a value that doesn't match.
</details>
Done.

---

*[storage/replica_raftstorage.go, line 280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRUTQCazzfm332k4CXU:-KRZ8lJxZMOpHw6Rm0y0:b-bcr7l9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica_raftstorage.go#L280)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> My concern here is just the indirect nature of the path from here to CloseOutSnap. If raft fetches a snapshot from the replica but then for any reason fails to send the MsgSnap, we'll never close the snapshot and be blocked from sending any more snapshots of this replica in the future. Currently, it doesn't look like that can ever happen, but it also doesn't look like a very strong guarantee. I'd like to have handleRaftReady close any outgoing snapshot (not just when an MsgSnap is sent) as a safety net. 
</details>
Okay, here's my understanding of the situation as it is for raft-initiated snapshots.

When we call raft.Step in `processRaftRequest`, Raft may request an outgoing snapshot from storage and ultimately cause a call to `SnapshotWithContext`. The outgoing snapshot will then sit around on the `Replica` until `processReady` is called by the scheduler, triggering `handleRaftReady`, which eventually processes all of the outgoing Ready messages, triggering `sendRaftMessage` for each.

At the moment, `sendRaftMessage` is ultimately responsible for either calling `CloseOutSnap` or making sure that the snapshot stream goroutine does so. There's currently an opportunity for escape of an un-closed OutgoingSnapshot in between the call to `raft.Step` and the triggering of `sendRaftMessage`.

So, I don't think that having `handleRaftReady` deal with any held-open outgoing snapshot is good enough either, is it? If we want to be precise, we should make sure to defer a close to the outgoing snapshot as soon as `processReady` starts, aborting the defer if the snapshot ever gets picked up by the stream goroutine... or something like that. It seems pretty complicated. I'm going to think on this a bit more.

---

*[storage/replica_raftstorage.go, line 353 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRXXbDCTGabKPLfWXC5:-KRZ8PjpzSrrdrY1kPTj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica_raftstorage.go#L353)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/Close/Replica.CloseOutSnap/
</details>
Done.

---

*[storage/store.go, line 2635 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRXUKPLviT9zqcXa3yj:-KRZ8fQ0h4NHcxGFVtZQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/store.go#L2635)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This looks like a commented-out `continue` statement, which wouldn't make since because there is no loop. Either remove it or make it a sentence.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 20 of 21 files at r4.
Review status: 23 of 24 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/replica_raftstorage.go, line 280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRUTQCazzfm332k4CXU:-KRi-4RB1HmeIeIpJJwM:b2fsq27) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica_raftstorage.go#L280)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> Okay, here's my understanding of the situation as it is for raft-initiated snapshots.
> 
> When we call raft.Step in `processRaftRequest`, Raft may request an outgoing snapshot from storage and ultimately cause a call to `SnapshotWithContext`. The outgoing snapshot will then sit around on the `Replica` until `processReady` is called by the scheduler, triggering `handleRaftReady`, which eventually processes all of the outgoing Ready messages, triggering `sendRaftMessage` for each.
> 
> At the moment, `sendRaftMessage` is ultimately responsible for either calling `CloseOutSnap` or making sure that the snapshot stream goroutine does so. There's currently an opportunity for escape of an un-closed OutgoingSnapshot in between the call to `raft.Step` and the triggering of `sendRaftMessage`.
> 
> So, I don't think that having `handleRaftReady` deal with any held-open outgoing snapshot is good enough either, is it? If we want to be precise, we should make sure to defer a close to the outgoing snapshot as soon as `processReady` starts, aborting the defer if the snapshot ever gets picked up by the stream goroutine... or something like that. It seems pretty complicated. I'm going to think on this a bit more.
</details>
To be specific, my concern is that if raft were smarter, it could abandon snapshots when messages arrive in weird sequences. Suppose we get an old MsgAppResp which tells raft the follower needs a snapshot, followed by a new one that says it's caught up. Raft could have requested a snapshot after the first one and then dropped the outgoing message when processing the second one.

If that were to happen, we could detect it in `handleRaftReady`, since that's when we either get the outgoing raftpb.Message to send or not. Without quiescent ranges, that's good enough: ticks guarantee that `handleRaftReady` will be called at reasonable intervals. With quiescent ranges, it does get a little tricker since we need to ensure that we get at least one more ready cycle. Maybe we just add a check for pending snapshots in the quiescing logic.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I've got to do another on this later today.

---

Reviewed 3 of 3 files at r1.
Review status: 23 of 24 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1686 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KR_KamZ0zAGjx5mmH55:-KR_KamZ0zAGjx5mmH56:b-yoapl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica.go#L1686)):*
> ```Go
> // handleRaftReadyLocked is the same as handleRaftReady but requires that the
> // replica be locked for raft processing via r.raftLock.
> func (r *Replica) handleRaftReadyLocked(uninitRaftLocked *bool, inSnap IncomingSnapshot) error {
> ```

The convention we've adopted is that a `Locked` suffix indicates `Replica.mu` is locked by the caller and `RaftMuLocked` indicates that `Replica.raftMu` is locked by the caller. This method should be named `handleRaftReadyRaftMuLocked`. Bit of a mouthful, but consistency.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-:-KRiC_fz241pxTZqcaav:b429bap)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 19 of 24 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1686 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KR_KamZ0zAGjx5mmH55:-KRiJEaZlK51a_U6_x2s:bqvz78q) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica.go#L1686)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The convention we've adopted is that a `Locked` suffix indicates `Replica.mu` is locked by the caller and `RaftMuLocked` indicates that `Replica.raftMu` is locked by the caller. This method should be named `handleRaftReadyRaftMuLocked`. Bit of a mouthful, but consistency.
</details>
Ok, done. Paging Dr. Seuss... 
When beetles fight these battles in a bottle with their paddles 
and the bottle's on a poodle and the poodle's eating noodles...
...they call this a muddle puddle tweetle poodle beetle noodle 
bottle paddle battle.

---

*[storage/replica_raftstorage.go, line 280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRUTQCazzfm332k4CXU:-KRiQYml7lAvXof6UsM3:b499hrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica_raftstorage.go#L280)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> To be specific, my concern is that if raft were smarter, it could abandon snapshots when messages arrive in weird sequences. Suppose we get an old MsgAppResp which tells raft the follower needs a snapshot, followed by a new one that says it's caught up. Raft could have requested a snapshot after the first one and then dropped the outgoing message when processing the second one.
> 
> If that were to happen, we could detect it in `handleRaftReady`, since that's when we either get the outgoing raftpb.Message to send or not. Without quiescent ranges, that's good enough: ticks guarantee that `handleRaftReady` will be called at reasonable intervals. With quiescent ranges, it does get a little tricker since we need to ensure that we get at least one more ready cycle. Maybe we just add a check for pending snapshots in the quiescing logic.
</details>
I addressed this in `handleRaftReady` for now. I guess I should also add a similar check in the quiescing code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
A few comments on the interaction between `HandleSnapshot`, `HandleRaftRequest` and `processRaftRequest`. It's possible I'm leading you astray so don't hesitate to push-back if I'm missing some detail of the current organization.

---

Review status: 19 of 24 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/replica_raftstorage.go, line 280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRUTQCazzfm332k4CXU:-KRjTRMLIih0vyjoZ9Mx:b-ux1g7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica_raftstorage.go#L280)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> I addressed this in `handleRaftReady` for now. I guess I should also add a similar check in the quiescing code.
</details>
Yeah, I think we shouldn't quiesce if there is a pending snapshot. See `maybeQuiesceLocked`. 

---

*[storage/store.go, line 2286 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjTw6bNVq-FyXBXHMx:-KRjTw6bNVq-FyXBXHMy:b-numljj) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2286)):*
> ```Go
> 	stream MultiRaft_RaftSnapshotServer,
> ) error {
> ```

Spurious blank line.

---

*[storage/store.go, line 2296 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjVttyQAYfnKTPVATt:-KRjVttyQAYfnKTPVATu:bcs8wz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2296)):*
> ```Go
> 	ctx := s.logContext(stream.Context())
> 
> 	var addedPlaceholder bool
> ```

This code moved here because you aren't passing the snapshot stream through to `processRaftRequest`. If you passed the stream through to `processRaftRequest` this code could move back there where I think it more naturally belongs. You'd also get rid of `IncomingSnapshot.{added,removed}Placeholder`.

---

*[storage/store.go, line 2374 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjUvSlhV_npj-UY1NM:-KRjUvSlhV_npj-UY1NN:b-99kk9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2374)):*
> ```Go
> 			}
> 
> 			if err := s.handleRaftRequest(ctx, &header.RaftMessageRequest, inSnap, nil); err != nil {
> ```

Add a `TODO(jordan)` that we should be passing the stream through to `handleRaftRequest` and not reading from it until it is time to process the snapshot. 

Probably should be calling `processRaftRequest` directly here. The indirection through `handleRaftRequest` isn't terribly useful other than incrementing the stats and that should be done at the start of this method. This would also get rid of the addition of `handleRaftRequest`.

---

*[storage/store.go, line 2609 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjXGpXkEbNM9c2SuEQ:-KRjXGpYglhXhdTF6ZuV:bybz4av) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2609)):*
> ```Go
> 
> 	if (inSnap.SnapUUID == uuid.UUID{}) {
> 		s.enqueueRaftUpdateCheck(req.RangeID)
> ```

We could actually always calling `handleRaftReadyRaftMuLocked` instead of enqueuing an update check which will end up doing the same. Mind adding a `TODO(peter)` to investigate that?

---

*[storage/store.go, line 2672 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjW963qMwkCBxRNGHE:-KRjW963qMwkCBxRNGHF:b-numljj) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2672)):*
> ```Go
> 	newBatch func() engine.Batch,
> ) error {
> ```

Spurious blank line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-:-KRjXVOrpLa87UUAEeII:bfv7kz)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTR!

---

Review status: 18 of 24 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/replica_raftstorage.go, line 280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRUTQCazzfm332k4CXU:-KRjZyYPJrc9QSptKZgy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4999162bdc5f8dbfd62162579b2ac93745157a80/storage/replica_raftstorage.go#L280)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yeah, I think we shouldn't quiesce if there is a pending snapshot. See `maybeQuiesceLocked`. 
</details>
Done.

---

*[storage/store.go, line 2286 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjTw6bNVq-FyXBXHMx:-KRjY1D5wlk7tGm_-aAx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2286)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Spurious blank line.
</details>
Done.

---

*[storage/store.go, line 2296 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjVttyQAYfnKTPVATt:-KRjaVslnc08ZoyEfukX:bl5f1ln) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2296)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This code moved here because you aren't passing the snapshot stream through to `processRaftRequest`. If you passed the stream through to `processRaftRequest` this code could move back there where I think it more naturally belongs. You'd also get rid of `IncomingSnapshot.{added,removed}Placeholder`.
</details>
Hmm, I'm not convinced that this code more naturally belongs in `processRaftRequest`.

1. The stream processing/validation steps don't need to happen under the stricter lock that `processRaftRequest` holds. There's a bunch of negotiation here that can happen before we even consider beginning to process the snapshot. Additionally, in the future we may choose to implement a flow control mechanism for snapshots (see the RFC) that works by holding the stream open until a condition is met. That would certainly need to happen outside of the raft lock.

2. `processRaftRequest` needs to be able to deal with snapshot and non-snapshot requests. Moving the snapshot-only stream negotiation code into `processRaftRequest` seems confusing to me - that function is already pretty complicated, and I think it would get significantly worse with this code move.

---

*[storage/store.go, line 2374 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjUvSlhV_npj-UY1NM:-KRjaTXm381Ik8E_BCEX:b-eoolgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2374)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Add a `TODO(jordan)` that we should be passing the stream through to `handleRaftRequest` and not reading from it until it is time to process the snapshot. 
> 
> Probably should be calling `processRaftRequest` directly here. The indirection through `handleRaftRequest` isn't terribly useful other than incrementing the stats and that should be done at the start of this method. This would also get rid of the addition of `handleRaftRequest`.
</details>
Done, removed `handleRaftRequest`.

---

*[storage/store.go, line 2609 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjXGpXkEbNM9c2SuEQ:-KRjYWd0pJWdYbNYIRhf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2609)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> We could actually always calling `handleRaftReadyRaftMuLocked` instead of enqueuing an update check which will end up doing the same. Mind adding a `TODO(peter)` to investigate that?
</details>
Done.

---

*[storage/store.go, line 2672 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjW963qMwkCBxRNGHE:-KRjYK_AMcPa1vN3f8al:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2672)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Spurious blank line.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-:-KRjjimzX2PAG5Na-LSU:b-814g8k)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Reviewed 21 of 21 files at r3.
Review status: 18 of 24 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 2296 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjVttyQAYfnKTPVATt:-KRjklcia97jY8PnxtP9:bsgcn6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2296)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> Hmm, I'm not convinced that this code more naturally belongs in `processRaftRequest`.
> 
> 1. The stream processing/validation steps don't need to happen under the stricter lock that `processRaftRequest` holds. There's a bunch of negotiation here that can happen before we even consider beginning to process the snapshot. Additionally, in the future we may choose to implement a flow control mechanism for snapshots (see the RFC) that works by holding the stream open until a condition is met. That would certainly need to happen outside of the raft lock.
> 
> 2. `processRaftRequest` needs to be able to deal with snapshot and non-snapshot requests. Moving the snapshot-only stream negotiation code into `processRaftRequest` seems confusing to me - that function is already pretty complicated, and I think it would get significantly worse with this code move.
</details>
The separation of the placeholder addition here and removal in `processRaftRequest` is a bit ugly, IMO. The stream processing/validation steps could be encapsulated in a `processSnapshotStream` method which returns an `IncomingSnapshot`. Part of my reason for suggesting this is that it would reduce the code movement in this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r1, 16 of 21 files at r4, 2 of 4 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 603 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjpM00n8r6zml7mhvj:-KRjpM01C8ds0Vnrq5MI:b-edx74i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b8e139e2741a7f1f3f4173ea6ffd396580700ce/storage/raft_transport.go#L603)):*
> ```Go
> 		defer func() {
> 			if err := stream.CloseSend(); err != nil {
> 				log.Warningf(stream.Context(), "failed to close snapshot stream: %s", err)
> ```

i think this should use `ctx`, not stream.Context()

---

*[storage/store_test.go, line 2444 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjosmIo3TbhupzUS_t:-KRjosmIo3TbhupzUS_u:b-p1zra1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b8e139e2741a7f1f3f4173ea6ffd396580700ce/storage/store_test.go#L2444)):*
> ```Go
> 	placeholder, err := s.canApplySnapshotLocked(rng1.Desc())
> 	if err != nil {
> 		t.Fatal(err)
> ```

if this hits you won't unlock the store lock, and likely cause the test to deadlock on shutdown.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

tamird commented:
You probably have a merge conflict + merge skew now that @raduberinde removed log.Trace{,f}

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-:-KRjppIOMBhTbmp44DqC:b-9sp6lo)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/replica_raftstorage.go, line 297 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjto_hBqvlCIt6GZOE:-KRjto_hBqvlCIt6GZOF:b3s3tmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b8e139e2741a7f1f3f4173ea6ffd396580700ce/storage/replica_raftstorage.go#L297)):*
> ```Go
> 	case <-r.mu.outSnapDone:
> 	default:
> 		log.Trace(ctx, "snapshot already running")
> ```

These  will just become `log.Event`/`Eventf`.  You'll probably see conflicts for the pieces you deleted that include `Trace` but it's just because of that rename. Hope it's not too much trouble.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Yep, just fixed all of those.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 603 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjpM00n8r6zml7mhvj:-KRjqmMk6gg6qBkuU4bi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b8e139e2741a7f1f3f4173ea6ffd396580700ce/storage/raft_transport.go#L603)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i think this should use `ctx`, not stream.Context()
</details>
Done.

---

*[storage/replica_raftstorage.go, line 297 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjto_hBqvlCIt6GZOE:-KRjuA4Nn4mwYKWQCq8D:b94tp1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b8e139e2741a7f1f3f4173ea6ffd396580700ce/storage/replica_raftstorage.go#L297)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> These  will just become `log.Event`/`Eventf`.  You'll probably see conflicts for the pieces you deleted that include `Trace` but it's just because of that rename. Hope it's not too much trouble.
</details>
Yeah no biggie, done.

---

*[storage/store.go, line 2296 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjVttyQAYfnKTPVATt:-KRjuHe91hnQtL7EQCH4:b-mx1hbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2296)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The separation of the placeholder addition here and removal in `processRaftRequest` is a bit ugly, IMO. The stream processing/validation steps could be encapsulated in a `processSnapshotStream` method which returns an `IncomingSnapshot`. Part of my reason for suggesting this is that it would reduce the code movement in this PR.
</details>
Okay I'll try that out next.

---

*[storage/store_test.go, line 2444 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjosmIo3TbhupzUS_t:-KRjreM1Kr18x4k2Yo17:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b8e139e2741a7f1f3f4173ea6ffd396580700ce/storage/store_test.go#L2444)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> if this hits you won't unlock the store lock, and likely cause the test to deadlock on shutdown.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-:-KRjrgHWMxJOUv6jvf4a:b-txreoy)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 21 of 21 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/store_test.go, line 2475 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjxnKKxg24MQZecVCd:-KRjxnKKxg24MQZecVCe:b-7lj86h) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d0e94271e4cd4ca5659ca6b353d3f66006a4f0a/storage/store_test.go#L2475)):*
> ```Go
> 	if n := atomic.LoadInt32(&s.counts.droppedPlaceholders); n != 1 {
> 		t.Errorf("expected 1 dropped placeholder, but found %d", n)
> ```

oops?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 21 of 21 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 2296 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjVttyQAYfnKTPVATt:-KRo5D3mzbbbxTKCONKc:bbrxevo) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2296)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> Okay I'll try that out next.
</details>
I've played around with this and I don't really see a good way of making this happen without mucking up the nice lock granularity that already exists. Do you see a clean way of implementing your suggestion without needing to do network IO under the raft locks? I agree that the separation of the placeholder addition is ugly but I think that to make this better I'd need to do a larger refactor that I'd rather not insert into this PR.

---

*[storage/store_test.go, line 2475 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjxnKKxg24MQZecVCd:-KRnaNNV0fMtK6qv_5lH:b-ph8d4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d0e94271e4cd4ca5659ca6b353d3f66006a4f0a/storage/store_test.go#L2475)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> oops?
</details>
Blank line? Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Reviewed 21 of 21 files at r7.
Review status: 3 of 24 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2296 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjVttyQAYfnKTPVATt:-KRooz5rEZHBbQh0RZCf:b-993w9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2296)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> I've played around with this and I don't really see a good way of making this happen without mucking up the nice lock granularity that already exists. Do you see a clean way of implementing your suggestion without needing to do network IO under the raft locks? I agree that the separation of the placeholder addition is ugly but I think that to make this better I'd need to do a larger refactor that I'd rather not insert into this PR.
</details>
I think we need to figure this out because I just realized your current code isn't valid. You have to hold the Replica's raftMu while inserting the placeholder or it can get removed out from under you. Reading the stream while holding `Replica.raftMu` isn't that bad. It only blocks up the one Replica and that's ok because we're applying either a preemptive or normal snapshot. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 24 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2296 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjVttyQAYfnKTPVATt:-KRoq1RecdY127WxaZ7T:b-btfhhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2296)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I think we need to figure this out because I just realized your current code isn't valid. You have to hold the Replica's raftMu while inserting the placeholder or it can get removed out from under you. Reading the stream while holding `Replica.raftMu` isn't that bad. It only blocks up the one Replica and that's ok because we're applying either a preemptive or normal snapshot. 
</details>
And please no large refactor. My idea is that `processRaftRequest` should look very similar to the way it does before this PR but with the addition that it has to read the snapshot from the stream (if there is a snapshot).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 3 of 24 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2296 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjVttyQAYfnKTPVATt:-KRp3Em89w1D8lJrCPe2:b2c5dix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2296)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> And please no large refactor. My idea is that `processRaftRequest` should look very similar to the way it does before this PR but with the addition that it has to read the snapshot from the stream (if there is a snapshot).
</details>
I gave that a shot - I don't like having `processRaftRequest` require taking a GRPC stream. It makes testing that function much more difficult.

Tobi gave me the idea of having `HandleSnapshot` just do a best-effort check to see if the snapshot is applyable and not actually add a placeholder. Then it will finish downloading the snapshot in full and send it to `processRaftRequest` which will then re-check the validity of the snapshot and add the placeholder if necessary, which restores the behavior of the function from before. Sound good? I'm going to upload a patch that does that soon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 24 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2296 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjVttyQAYfnKTPVATt:-KRpYPUofZVtG8PFuT_Q:brhzzyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2296)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> I gave that a shot - I don't like having `processRaftRequest` require taking a GRPC stream. It makes testing that function much more difficult.
> 
> Tobi gave me the idea of having `HandleSnapshot` just do a best-effort check to see if the snapshot is applyable and not actually add a placeholder. Then it will finish downloading the snapshot in full and send it to `processRaftRequest` which will then re-check the validity of the snapshot and add the placeholder if necessary, which restores the behavior of the function from before. Sound good? I'm going to upload a patch that does that soon.
</details>
That sounds ok, though I think encapsulating the stream in a type could avoid the ugliness of passing a stream to `processRaftRequest`. The important part to me is that `processRaftRequest` adds the placeholder after locking `Replica.raftMu`. In the future we'll want to process the stream directly into a C++ WriteBatch which may make not passing the stream awkward. Maybe. I haven't fully thought that through yet.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 3 of 24 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2296 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRjVttyQAYfnKTPVATt:-KRqN-4La6i81I0xZQgq:b-3377wl) ([raw file](https://github.com/cockroachdb/cockroach/blob/be93f7cea5228a901757e7438c27d4c920cac1d6/storage/store.go#L2296)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> That sounds ok, though I think encapsulating the stream in a type could avoid the ugliness of passing a stream to `processRaftRequest`. The important part to me is that `processRaftRequest` adds the placeholder after locking `Replica.raftMu`. In the future we'll want to process the stream directly into a C++ WriteBatch which may make not passing the stream awkward. Maybe. I haven't fully thought that through yet.
</details>
For now I've gone with the simple solution above. Let's keep the wrapped stream in mind for a future PR, especially since eventually we'd like to rework this to get true streaming on the recipient side.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 21 of 21 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2321 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRt1JGliesyOHnkMbVk:-KRt1JGliesyOHnkMbVl:b-ezylqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab06f2281a34ec15b9a25be55a4b67c89d33bb70/storage/store.go#L2321)):*
> ```Go
> 	// a placeholder here. We'll recheck later.
> 	s.mu.Lock()
> 	_, err := s.canApplySnapshotLocked(&header.RangeDescriptor)
> ```

is there a `canApplySnapshot`?

---

*[storage/store_test.go, line 2349 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRt0ltu6zzHWE3fVYyR:-KRt0ltu6zzHWE3fVYyS:b-rl7xpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab06f2281a34ec15b9a25be55a4b67c89d33bb70/storage/store_test.go#L2349)):*
> ```Go
> 			RangeDescriptor: *rng1.Desc(),
> 		}); !testutils.IsError(
> 		errors.Errorf("%s", err), expected) {
> ```

`errors.Errorf("%s", err)` wat

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2319 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRt6X8SNtYl6AxhG7hm:-KRt6X8SNtYl6AxhG7hn:bt1n3ur) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab06f2281a34ec15b9a25be55a4b67c89d33bb70/storage/store.go#L2319)):*
> ```Go
> 
> 	// Check to see if the snapshot can be applied but don't attempt to add
> 	// a placeholder here. We'll recheck later.
> ```

You should expand this to note why a placeholder can't be added here (we're not holding `Replica.raftMu`) and that this is purely an optimization to avoid reading the snapshot if it can't be applied. 

---

*[storage/store.go, line 2321 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRt1JGliesyOHnkMbVk:-KRt6nuiLME-EyC5G4TE:bojb4lj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab06f2281a34ec15b9a25be55a4b67c89d33bb70/storage/store.go#L2321)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is there a `canApplySnapshot`?
</details>
We should add one.

---

*[storage/store.go, line 2325 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRt6rqHPyZV5gFS1bXu:-KRt6rqHPyZV5gFS1bXv:b-yc4uc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab06f2281a34ec15b9a25be55a4b67c89d33bb70/storage/store.go#L2325)):*
> ```Go
> 	if err != nil {
> 		return sendSnapError(
> 			errors.Wrapf(err, "%s: cannot apply snapshot on range %d", s, header.RangeDescriptor.RangeID),
> ```

It is nice to have every log message format range IDs as `r%d`. That is the way they appear in `Replica.ctx` and `Replica.String()`. The consistency makes for better grepability. For this message I'd suggest: 

```
errors.Wrapf(err, "%s,r%d: cannot apply snapshot", s, header.RangeDescriptor.RangeID)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-:-KRt7spo8GBaukO2hDoY:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 22 of 24 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2319 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRt6X8SNtYl6AxhG7hm:-KRykI5hN4bRye9iQB4J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab06f2281a34ec15b9a25be55a4b67c89d33bb70/storage/store.go#L2319)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> You should expand this to note why a placeholder can't be added here (we're not holding `Replica.raftMu`) and that this is purely an optimization to avoid reading the snapshot if it can't be applied. 
</details>
Done.

---

*[storage/store.go, line 2321 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRt1JGliesyOHnkMbVk:-KRyktnt5OC617BcvAWE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab06f2281a34ec15b9a25be55a4b67c89d33bb70/storage/store.go#L2321)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> We should add one.
</details>
Done.

---

*[storage/store.go, line 2325 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRt6rqHPyZV5gFS1bXu:-KRyl29O8TUFA5q21dwH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab06f2281a34ec15b9a25be55a4b67c89d33bb70/storage/store.go#L2325)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> It is nice to have every log message format range IDs as `r%d`. That is the way they appear in `Replica.ctx` and `Replica.String()`. The consistency makes for better grepability. For this message I'd suggest: 
> 
> ```
> errors.Wrapf(err, "%s,r%d: cannot apply snapshot", s, header.RangeDescriptor.RangeID)
> ```
</details>
Done.

---

*[storage/store_test.go, line 2349 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRt0ltu6zzHWE3fVYyR:-KRymIh31U0og0A5TtYl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab06f2281a34ec15b9a25be55a4b67c89d33bb70/storage/store_test.go#L2349)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `errors.Errorf("%s", err)` wat
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/store_test.go, line 2349 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRt0ltu6zzHWE3fVYyR:-KRzqrTKHWY21W990LQP:bp71frm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab06f2281a34ec15b9a25be55a4b67c89d33bb70/storage/store_test.go#L2349)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> Done.
</details>
You can also use `testutils.IsPError`, I think.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 3 of 24 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/store_test.go, line 2349 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292#-KRt0ltu6zzHWE3fVYyR:-KSCBgXxNK4oFOzrIc3N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab06f2281a34ec15b9a25be55a4b67c89d33bb70/storage/store_test.go#L2349)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> You can also use `testutils.IsPError`, I think.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 21 of 21 files at r13.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9292)*
<!-- Sent from Reviewable.io -->
